### ISSUE TITLE
[DataGrid] Hidden columns widths still used in calculations

### ISSUE THREAD

**Issue by jzecca** [mui-org/material-ui-x#708]
- [X] The issue is present in the latest release.
- [X] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

Since 4.0.0-alpha.10, a column with `flex: 1` fills the available space.

This works well, until you hide one of the columns using `hide: true`:  
Then the flex column width remains the same, as if the hidden column was still there.

_(sorry if I'm not clear enough, the codesanbox below should be self-explanatory though 😉)_

## Expected Behavior 🤔

The flex column should adjust to always take up all available space.
Hidden columns should be ignored in width calculations.

## Steps to Reproduce 🕹

https://codesandbox.io/s/mui-datagrid-width-issue-with-hidden-columns-vx8ny?file=/demo.js

Steps A:

1. The lastname column takes all the remaining space, thanks to `flex: 1`. Everything's OK.
2. Click _"Hide "First name" column"_
3. The width of the lastname column does not adjust.

Steps B:

1. Set default visibility to false on L17 & refresh
2. Same issue with the column initialy hidden.

Steps C:

1. Remove the `firstName` column completely (L23 to 28)
2. The flex column behaves as expected.

## Your Environment 🌎

| Tech          | Version |
| ------------- | ------- |
| Material-UI X | v4.0.0-alpha.11  |
| React         |  v17.0.1       |
| Browser       |  Chrome       |

**Comment by DanailH** [comment]
@jzecca Thanks for reporting this. I'll have a look.

**Comment by willenglishiv** [comment]
+1 to this issue. I try to hide a column (to then use it with another column), the flex column not only exists but then resizes itself (max/min width)

**Comment by oliviertassinari** [comment]
A simplified reproduction https://codesandbox.io/s/mui-datagrid-width-issue-with-hidden-columns-forked-vpf6b?file=/demo.js

```jsx
import * as React from "react";
import { DataGrid } from "@material-ui/data-grid";

const rows = [{ id: 1, lastName: "Snow", firstName: "Jon", age: 35 }];

export default function DataGridDemo() {
  const columns = [
    { field: "id", headerName: "ID", flex: 1 },
    {
      field: "firstName",
      headerName: "First name",
      hide: true
    }
  ];

  return (
    <div style={{ height: 400, width: "100%" }}>
      <DataGrid rows={rows} columns={columns} pageSize={5} />
    </div>
  );
}
```

<img width="518" alt="Capture d’écran 2020-12-19 à 13 08 14" src="https://user-images.githubusercontent.com/3165635/102689013-4b3fc800-41fb-11eb-8af6-aeb1739de988.png">

**Comment by dmitryKochergin** [comment]
> A simplified reproduction https://codesandbox.io/s/mui-datagrid-width-issue-with-hidden-columns-forked-vpf6b?file=/demo.js
> 
> ```js-jsx
> import * as React from "react";
> import { DataGrid } from "@material-ui/data-grid";
> 
> const rows = [{ id: 1, firstName: "Jon", age: 35 }];
> const columns = [
>   { field: "id", headerName: "ID", flex: 1 },
>   {
>     field: "firstName",
>     headerName: "First name",
>     hide: true
>   }
> ];
> 
> export default function DataGridDemo() {
>   return (
>     <div style={{ height: 400, width: "100%" }}>
>       <DataGrid rows={rows} columns={columns} pageSize={5} />
>     </div>
>   );
> }
> ```
> 
> <img alt="Capture d’écran 2020-12-19 à 13 08 14" width="518" src="https://user-images.githubusercontent.com/3165635/102689013-4b3fc800-41fb-11eb-8af6-aeb1739de988.png">

I made some changes in this code and got a strange result:
https://codesandbox.io/s/mui-datagrid-width-issue-with-hidden-columns-forked-3i4tk?file=/demo.js

After clicking the button, all columns changes its width to default (100px, even if they had a flex property).
Should I make an issue about that?

**Comment by oliviertassinari** [comment]
@dmitryKochergin This is a difference issue, and actually a regression, see #716.

**Comment by si7o** [comment]
Hi @oliviertassinari I was checking the the simplified reproduction you posted and noticed something weird (sorry if there is an issue for this already, I could not find it):

Using DataGrid 4.0.0-alpha.31

https://user-images.githubusercontent.com/5405326/123953124-2c3ea180-d9a7-11eb-917e-154e26012cd8.mp4

I changed the column visibility and, instead of resizing the 2 columns, an horizontal scrollbar appeared in the DataGrid, then the columns wont resize until I change the browser/window size manually.

This happened in firefox & edge browser. On Chrome it worked slightly better, but when hiding a column I still had to resize the window so the DataGrid columns resize.

I found this issue because I have a DataGrid with several columns and they dont resize when I hide some of them
